[SPARK-26718][SS][BRANCH-2.4] Fixed integer overflow in SS kafka rateLimit calculation

linehrr opened the pull request:
## What changes were proposed in this pull request?

Fix the integer overflow issue in rateLimit.

## How was this patch tested?

Pass the Jenkins with newly added UT for the possible case where integer could be overflowed.

Commit:
fixed integer overflow: https://issues.apache.org/jira/browse/SPARK-26718

 On branch fix/integer_overflow_rateLimit
 Changes to be committed:
	modified:   external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchReader.scala

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

gaborgsomogyi commented:
Please check commit message requirements + I think this can be unit tested.

dongjoon-hyun commented:
Hi, @linehrr . Thank you for your first contribution.
Could you make a PR to `master` branch first? We usually merge into `master` first for the future release and do backport from it.

Commit:
added unit test for rateLimit integer overflow

 On branch fix/integer_overflow_rateLimit
 Your branch is up to date with 'origin/fix/integer_overflow_rateLimit'.

 Changes to be committed:
	modified:   src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

linehrr commented:
@dongjoon-hyun 
Added unit test, it fails before the code fix change and succeeded after the fix. 
I can't put this PR against the master, since master does not even have this class, maybe master is too behind or too ahead.

dongjoon-hyun commented:
Thank you for the UT, @linehrr .
And, do you mean your UT pass in the master?
For the master, the patch might be different.

linehrr commented:
@dongjoon-hyun 
in master, the whole class `sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchReader.scala` is not even there.

dongjoon-hyun commented:
@linehrr .
Do you want to find this? You can use `git grep` to find the line you need.
- https://github.com/apache/spark/blob/master/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala#L234

linehrr commented:
@dongjoon-hyun oh that's neat. I thought the whole class got deprecated in master, but it's just got renamed. I will do a PR in master as well. but do we want to backport this fix to 2.4.0 branch as well ?

dongjoon-hyun commented:
For `branch-2.4`, you can keep this PR since the patch is not the same.

dongjoon-hyun commented:
BTW, if you do `git grep`, there are one additional instance like this. Please check that together.

linehrr commented:
@dongjoon-hyun hmm.. looks like KafkaSource.scala has the exact same code there... 
this is not good, we have duplicated code. and also I can't find good place like shared base trait or class to abstract that into a shared function... for now, I will just fix 2 places.

Commit:
fixed another place with the same code

 On branch fix/integer_overflow_rateLimit
 Your branch is up to date with 'origin/fix/integer_overflow_rateLimit'.

 Changes to be committed:
	modified:   external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala

srowen reviewed (commented):
Seems OK to me, just open a PR for master first, yeah

HeartSaVioR commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:
```diff
@@ -1057,6 +1092,7 @@ class KafkaMicroBatchV2SourceSuite extends KafkaMicroBatchSourceSuiteBase {
     intercept[IllegalArgumentException] { test(minPartitions = "-1", 1, true) }
   }
 
+
```
nit: let's remove unnecessary change.

dongjoon-hyun replied:
So, as @HeartSaVioR suggested, we should not add this line.

HeartSaVioR commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:
```diff
@@ -199,6 +199,41 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
   }
 
+
+  test("Rate limit set to Long.Max should not overflow integer during end offset calculation[SPARK-26718]") {
```
nit: There's actually no rule on this, but most of the cases JIRA title is used as a prefix (search "SPARK-" against `*Suite.scala`) so might be better to follow the case.

HeartSaVioR reviewed (commented):
LGTM except some minor nits. DISCLAIMER: I suggested the logic so my review is just around new test case.

HeartSaVioR commented:
Btw, I guess this patch add a new UT so please update the content of PR. 

And we may want to get some information on commit message: requiring to open JIRA for details might be annoying if we need to deal with multiple commits for triaging some bugs. Don't need to contain all the thing, but IMHO the mandatory information would be 1. the bug is related to Kafka source in spark-sql-kafka 2. the patch fixes the integer overflow issue in rateLimit. Current title and content of PR are not clear where the bug is.

linehrr commented:
submitted this PR for master branch here: https://github.com/apache/spark/pull/23666, changed a bit on the PR title and reverted some unnecessary changes mentioned by @HeartSaVioR.

dongjoon-hyun commented:
Please update this PR like the patch at `master` branch.

dongjoon-hyun commented:
BTW, I just noticed that your commit authorship is unmatched to GitHub info.
```
ryne.yang <ryne.yang@acuityads.com>
```

It's not a problem, but you may want to be visible as GitHub ID here. :)
- https://github.com/apache/spark/pulse/daily

Commit:
fixed test case, improved test speed

 Changes to be committed:
	modified:   src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchReader.scala
	modified:   src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala
	modified:   src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

linehrr commented:
@dongjoon-hyun 
done. 
my github username got messed up by inheriting the global username set in my company laptop LOL. it's ok, no need to fix it.

dongjoon-hyun commented:
ok to test

SparkQA commented:
**[Test build #101838 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101838/testReport)** for PR 23652 at commit [`a32718c`](https://github.com/apache/spark/commit/a32718c966785e9cf092a334f08697d5f16fe7cb).

dongjoon-hyun commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala:
```diff
@@ -188,9 +188,15 @@ private[kafka010] class KafkaSource(
             val begin = from.get(tp).getOrElse(fromNew(tp))
             val prorate = limit * (size / total)
-            logDebug(s"rateLimit $tp prorated amount is $prorate")
-            // Don't completely starve small topicpartitions
```
In the master, we didn't touch line 191 and 192.

dongjoon-hyun commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:
```diff
@@ -199,6 +199,42 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
       StopStream)
   }
 
+
```
Let's remove one extra empty line here. In master, we don't have this.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7539/
Test PASSed.

SparkQA commented:
**[Test build #101839 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101839/testReport)** for PR 23652 at commit [`a32718c`](https://github.com/apache/spark/commit/a32718c966785e9cf092a334f08697d5f16fe7cb).

Commit:
removed useless new line

reverted deletion of logging
 Changes to be committed:
	modified:   src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala
	modified:   src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

Commit:
removed another useless new line

 Changes to be committed:
	modified:   src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7540/
Test PASSed.

SparkQA commented:
**[Test build #101840 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101840/testReport)** for PR 23652 at commit [`667dc30`](https://github.com/apache/spark/commit/667dc30860a8c86359b8b5037586d3a3d02b23fb).

dongjoon-hyun commented:
Also, ping @maropu since this is also related to `branch-2.3`.

SparkQA commented:
**[Test build #101838 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101838/testReport)** for PR 23652 at commit [`a32718c`](https://github.com/apache/spark/commit/a32718c966785e9cf092a334f08697d5f16fe7cb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101838/
Test PASSed.

SparkQA commented:
**[Test build #101839 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101839/testReport)** for PR 23652 at commit [`a32718c`](https://github.com/apache/spark/commit/a32718c966785e9cf092a334f08697d5f16fe7cb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101839/
Test PASSed.

SparkQA commented:
**[Test build #101840 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101840/testReport)** for PR 23652 at commit [`667dc30`](https://github.com/apache/spark/commit/667dc30860a8c86359b8b5037586d3a3d02b23fb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101840/
Test PASSed.

dongjoon-hyun reviewed (approved):
+1, LGTM. Thank you, all.
Merged to branch-2.4.

dongjoon-hyun commented:
@linehrr . Could you make a PR once more? I cannot merge this to `branch-2.3` because there is a conflict on `branch-2.3`, too.

linehrr commented:
no worries, I will make one for 2.3 soon.

linehrr commented:
@dongjoon-hyun seeing the same error mentioned here showing up in 2.3 when trying to build the project: http://apache-spark-developers-list.1001551.n3.nabble.com/VOTE-SPARK-2-3-2-RC3-td24421.html

```
[ERROR] missing or invalid dependency detected while loading class file 'MetricsSystem.class'.
Could not access term eclipse in package org,
because it (or its dependencies) are missing. Check your build definition for
missing or conflicting dependencies. (Re-run with `-Ylog-classpath` to see the problematic classpath.)
A full rebuild may help if 'MetricsSystem.class' was compiled against an incompatible version of org.
[ERROR] missing or invalid dependency detected while loading class file 'MetricsSystem.class'.
Could not access term jetty in value org.eclipse,
because it (or its dependencies) are missing. Check your build definition for
missing or conflicting dependencies. (Re-run with `-Ylog-classpath` to see the problematic classpath.)
A full rebuild may help if 'MetricsSystem.class' was compiled against an incompatible version of org.eclipse.
[WARNING] one warning found
[ERROR] two errors found
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  13.405 s
[INFO] Finished at: 2019-01-29T17:13:23-05:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:3.2.2:testCompile (scala-test-compile-first) on project spark-sql-kafka-0-10_2.11: Execution scala-test-compile-first of goal net.alchim31.maven:scala-maven-plugin:3.2.2:testCompile failed.: CompileFailed -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException

```

srowen commented:
I've seen that a few times in the past and never figured it out. Same with a few others here. I think it's a build problem somewhere, not a Spark thing. Just try a clean build.

maropu commented:
Thanks for this work, @linehrr! I'll wait for your pr in branch-2.3.

linehrr commented:
@srowen even after doing `mvn clean test` it's still showing up. any suggestions ?

srowen commented:
I forget what I had to do the few times this happened. It did seem to happen after switching branches. Maybe ensure you shut down zinc? or if that doesn't work, delete all the spark artifacts under ~/.m2/repository?

linehrr commented:
@srowen strange, I don't have zinc server running and even after I purged whole .m2/repository still the same. anyway, I submitted the PR, I will delegate the build and test to jenkins. patch is almost the same as other branches, so it should be fine.
